src/core/attributes.rs

```diff
@@ -121,17 +131,31 @@ pub struct ObserveWhen<'a, T, F> {
     operation: Arc<F>,
 }
 
+pub struct OnFlushCancel (Arc<Fn()>);
+
+impl Cancel for OnFlushCancel {
+    fn cancel(&self) {
+        (self.0)()
+    }
+}
+
 impl<'a, T, F> ObserveWhen<'a, T, F>
 where
     F: Fn(Instant) -> MetricValue + Send + Sync + 'static,
     T: InputScope + WithAttributes + Send + Sync,
 {
     /// Observe the metric's value upon flushing the scope.
-    pub fn on_flush(self) {
+    pub fn on_flush(self) -> OnFlushCancel {
         let gauge = self.metric;
+        let metric_id = gauge.metric_id().clone();
         let op = self.operation;
         write_lock!(self.target.get_attributes().flush_listeners)
-            .push(Arc::new(move |now| gauge.write(op(now), Labels::default())));
+            .insert(metric_id.clone(),
+                    Arc::new(move |now| gauge.write(op(now), Labels::default())));
+        let flush_listeners = self.target.get_attributes().flush_listeners.clone();
+        OnFlushCancel(Arc::new(move || {
+            write_lock!(flush_listeners).remove(&metric_id);
```

Comment:
There is a bug here. `flush_listeners` should contain also an unique ID of the observer callback. The ID can be based on memory address of the instance if possible or on a random number.

```rust
    // examples/observer.rs

    let cancel1 = metrics.observe(metrics.gauge("uptime"), |_| 1).on_flush();
    // `process.uptime 1` is reported.
    let cancel2 = metrics.observe(metrics.gauge("uptime"), |_| 2).on_flush();
    // `process.uptime 2` is reported, `1` stopped without any cancelling.

    cancel1.cancel();
    // Nothing is reported, `2` stopped without cancelling. The cancel was triggered by a different handle.
```

Reply:
Addressed in #68